providers/azure.go

```diff
@@ -335,6 +403,54 @@ func getEmailFromJSON(json *simplejson.Json) (string, error) {
 	return email, err
 }
 
+func (p *AzureProvider) getGroupsFromJSON(json *simplejson.Json) ([]string, error) {
+	var groups []string
+	for _, doc := range json.Get("value").MustArray() {
+		for k, v := range doc.(map[string]interface{}) {
+			if k == p.GroupField {
+				groups = append(groups, v.(string))
+			}
+		}
+	}
+	return groups, nil
+}
+
+func (p *AzureProvider) getGroupsFromGroupsAPI(ctx context.Context, accessToken string) ([]string, error) {
+	if accessToken == "" {
+		return nil, errors.New("missing access token")
+	}
+	var groupURL = p.GroupURL.String()
+	var groups []string
+
+	for groupURL != "" {
+		// logger.Printf("Calling Group API: %s", groupURL)
+		jsonRequest, err := requests.New(groupURL).
```

Comment:
Have you verified this works when you set your scopes to `<CLIENT_ID>/.default`? The audience you're generated for will change and I don't think graph will accept it.

Reply:
The client credentials flow requires that the `https://graph.microsoft.com/.default` scope is present, see https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-client-creds-grant-flow

Indeed, if setting `--resource=<CLIENT_ID>/.default` I'm getting
```
Error creating session during OAuth2 callback: unable to get groups: unexpected status "401": {"error":{"code":"InvalidAuthenticationToken","message":"Invalid x5t claim.","innerError":{"date":"2022-02-26T18:43:04","request-id":"xxx","client-request-id":"xxx"}}}
```
But this should be expected because only the access_token from the Microsoft graph scope is authorised to access Microsoft graph api.

Reply:
This is not ideal. Users should not be passing Microsoft Graph tokens to their apps since it does not have an audience for their resource server. You'd probably need to do token exchange here

Reply:
> This is not ideal. Users should not be passing Microsoft Graph tokens to their apps since it does not have an audience for their resource server. You'd probably need to do token exchange here.

Agree wtih you. I've completely refactored it so that it now supports the `--resource` flag in both V1 and V2. Microsoft graph token is treated separate now, without influencing the current flow.

> In v1 endpoints, the default audience is your client. In v2 it is graph?

In v2.0 tokens, this is always the client ID of the API, while in v1.0 tokens it can be the client ID or the resource URI used in the request, depending on how the client requested the token.